src/openfermion/utils/_sparse_tools.py

```diff
@@ -241,45 +284,21 @@ def matvec(vec):
         """matvec for the LinearOperator class.
 
         Args:
-          vec(numpy.ndarray): 1D numpy array.
+            vec(numpy.ndarray): 1D numpy array.
 
         Returns:
-          retvec(numpy.ndarray): same to the shape of input vec.
+            numpy.ndarray: with the same shape to that of vec.
         """
         if len(vec) != n_hilbert:
             raise ValueError('Invalid length of vector specified: %d != %d'
                              %(len(vec), n_hilbert))
 
-        retvec = numpy.zeros(vec.shape, dtype=complex)
-        # Loop through the terms.
-        for qubit_term in qubit_operator.terms:
-            vecs = [vec]
-            tensor_factor = 0
-            coefficient = qubit_operator.terms[qubit_term]
-
-            for pauli_operator in qubit_term:
-                # Split vector by half and half for each bit.
-                if pauli_operator[0] > tensor_factor:
-                    vecs = [v for iter_v in vecs for v in numpy.split(
-                        iter_v, 2 ** (pauli_operator[0] - tensor_factor))]
-
-                # Note that this is to make sure that XYZ operations always work
-                # on vector pairs.
-                vec_pairs = [numpy.split(v, 2) for v in vecs]
-
-                # There is an non-identity op here, transform the vector.
-                xyz = {
-                    'X' : lambda vps: [[vp[1], vp[0]] for vp in vps],
-                    'Y' : lambda vps: [[-1j * vp[1], 1j * vp[0]] for vp in vps],
-                    'Z' : lambda vps: [[vp[0], -vp[1]] for vp in vps],
-                }
-                vecs = [v for vp in xyz[pauli_operator[1]](vec_pairs)
-                        for v in vp]
-                tensor_factor = pauli_operator[0] + 1
-
-            # No need to check tensor_factor, i.e. to deal with bits left.
-            retvec += coefficient * numpy.concatenate(vecs)
-        return retvec
+        pool = multiprocessing.Pool(processes)
+        vecs = pool.map(term_on_vec, [(numpy.copy(vec), item[0], item[1])
```

Comment:
My main worry for this is that for large systems, this is going to carry a heavy memory footprint, which will represent a tradeoff between the decreased computation time.  This is because this will store a copy for every operator in the sum, which could be in the 10's of thousands. 

The ideal fix would be some kind of adaptive parallel map-reduce that works through the map reduce tree with fixed chunk sizes.  However in lieu of something that complex, I think we have to set a parallelism limit to how many of these we store and work through the vector in parallel chunks.  For example, compute on 10 in parallel, reduce in parallel to 1, compute on next 10 in parallel, reduce those 10 with first etc.